google/cloud/firestore_v1/bulk_writer.py

```diff
@@ -0,0 +1,537 @@
+            The client that created this BulkWriter.
+    """
+
+    batch_size: int = 20
+
+    def __init__(
+        self,
+        client: Optional[BaseClient] = None,
+        options: Optional["BulkWriterOptions"] = None,
+    ):
+        self._client = client
+        self._options = options or BulkWriterOptions()
+        self._send_mode = self._options.mode
+        # Redundantly set this variable for IDE type hints
+        self._batch: BulkWriteBatch = self._reset_batch()
+        self._queued_batches = collections.deque([])
+        self._is_open: bool = True
+        self._is_sending: bool = False
+
+        self._success_callback: Optional[
+            Callable[[BaseDocumentReference, WriteResult, "BulkWriter"], NoReturn]
+        ] = None
+        self._batch_callback: Optional[
+            Callable[[BulkWriteBatch, BatchWriteResponse, "BulkWriter"], NoReturn]
+        ] = None
+        self._error_callback: Optional[Callable] = None
+
+        self._in_flight_documents: int = 0
+        self._rate_limiter = RateLimiter(
+            initial_tokens=self._options.initial_ops_per_second,
+            global_max_tokens=self._options.max_ops_per_second,
+        )
+
+        # Keep track of progress as batches and write operations are completed
+        self._total_batches_sent: int = 0
+        self._total_write_operations: int = 0
+
+        self._ensure_executor()
+
+    def _reset_batch(self) -> BulkWriteBatch:
+        self._batch = BulkWriteBatch(self._client)
+        return self._batch
+
+    def _ensure_executor(self):
+        self._executor = (
+            getattr(self, "_executor", None) or self._instantiate_executor()
+        )
+
+    def _ensure_sending(self):
+        if not self._is_sending:
+            self._ensure_executor()
+            self._send_until_queue_is_empty()
+
+    def _instantiate_executor(self):
+        return concurrent.futures.ThreadPoolExecutor()
+
+    def flush(self):
+        """
+        Block until all pooled write operations are complete and then resume
+        accepting new write operations.
+        """
+        # Calling `flush` consecutively is a no-op.
+        if not self._executor:
+            return
+
+        if len(self._batch) > 0:
+            self._enqueue_current_batch()
+
+        while self._queued_batches:
+            self._ensure_sending()
+            time.sleep(0.1)
+
+        self._is_sending = False
+        self._executor.shutdown()
+        # Completely release this resource, allowing our sending methods to
+        # easily detect if `flush` has been called and we should re-instantiate
+        # the executor. The reason for this is that `flush` hangs until everything
+        # is sent (and calling `shutdown` is the easiest way to do that), yet,
+        # it should not completely end the life of this BulkWriter. That role
+        # is filled by the `close` method.
+        self._executor = None
```

Comment:
Note: This method grows moderately more complex after error handling / retries are included, as retries introduce the possibility of more operations being queued potentially long after `flush()` is called and you thought you'd queued the last writes.